File: android-uxsdk-beta/src/main/java/dji/ux/beta/widget/focusmode/FocusModeWidget.java
/*
 * Copyright (c) 2018-2019 DJI
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dji.ux.beta.widget.focusmode;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.style.ForegroundColorSpan;
import android.util.AttributeSet;
import android.util.Pair;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.Dimension;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import dji.common.camera.SettingsDefinitions.FocusMode;
import dji.thirdparty.io.reactivex.Flowable;
import dji.thirdparty.io.reactivex.disposables.Disposable;
import dji.ux.beta.R;
import dji.ux.beta.base.DJISDKModel;
import dji.ux.beta.base.FrameLayoutWidget;
import dji.ux.beta.base.GlobalPreferencesManager;
import dji.ux.beta.base.SchedulerProvider;
import dji.ux.beta.base.uxsdkkeys.ObservableInMemoryKeyedStore;
import dji.ux.beta.util.SettingDefinitions;

/**
 * Widget will display the current focus mode of aircraft camera.
 * - MF text highlighted (in green) indicates focus mode is Manual Focus.
 * - AF text highlighted (in green) indicates focus mode is Auto Focus.
 * - AFC text highlighted (in green) indicates focus mode is Auto Focus Continuous.
 * <p>
 * Interaction:
 * Tapping will toggle between AF and MF mode.
 */
public class FocusModeWidget extends FrameLayoutWidget implements OnClickListener {

    //region constants
    private static final String TAG = "FocusModeWidget";
    //endregion

    //region fields
    private FocusModeWidgetModel widgetModel;
    private TextView titleTextView;
    private int activeColor;
    private int inactiveColor;
    private SchedulerProvider schedulerProvider;
    //endregion

    //region lifecycle
    public FocusModeWidget(Context context) {
        super(context);
    }

    public FocusModeWidget(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public FocusModeWidget(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void initView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        inflate(context, R.layout.uxsdk_widget_focus_mode_switch, this);
        setBackgroundResource(R.drawable.uxsdk_background_black_rectangle);
        titleTextView = findViewById(R.id.text_view_camera_control_af);
        schedulerProvider = SchedulerProvider.getInstance();
        if (!isInEditMode()) {
            widgetModel = new FocusModeWidgetModel(DJISDKModel.getInstance(),
                    ObservableInMemoryKeyedStore.getInstance(),
                    GlobalPreferencesManager.getInstance(),
                    schedulerProvider);
        }
        setOnClickListener(this);
        activeColor = getResources().getColor(R.color.uxsdk_green);
        inactiveColor = getResources().getColor(R.color.uxsdk_white);
        if (attrs != null) {
            initAttributes(context, attrs);
        }
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (!isInEditMode()) {
            widgetModel.setup();
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        if (!isInEditMode()) {
            widgetModel.cleanup();
        }
        super.onDetachedFromWindow();
    }

    @Override
    protected void reactToModelChanges() {
        reactToFocusModeChange();
    }

    @NonNull
    @Override
    public String getIdealDimensionRatioString() {
        return getResources().getString(R.string.uxsdk_widget_default_ratio);
    }

    @Override
    public void onClick(View v) {
        addDisposable(widgetModel.toggleFocusMode().subscribe(() -> {
            // Do nothing
        }, logErrorConsumer(TAG, "switch focus mode: ")));
    }

    //endregion

    //region private helpers

    private void checkAndUpdateUI() {
        addDisposable(reactToFocusModeChange());
    }

    private Disposable reactToFocusModeChange() {
        return Flowable.combineLatest(widgetModel.isAFCEnabled(), widgetModel.getFocusMode(), Pair::new)
                .observeOn(schedulerProvider.ui())
                .subscribe(values -> updateUI(values.first, values.second),
                        logErrorConsumer(TAG, "react to Focus Mode Change: "));
    }

    private void updateUI(boolean isAFCEnabled, FocusMode focusMode) {
        int autoFocusTextColor;
        int manualFocusTextColor;
        if (focusMode == FocusMode.MANUAL) {
            manualFocusTextColor = activeColor;
            autoFocusTextColor = inactiveColor;
        } else {
            autoFocusTextColor = activeColor;
            manualFocusTextColor = inactiveColor;
        }

        String autoFocusText;
        if (isAFCEnabled) {
            autoFocusText = getResources().getString(R.string.uxsdk_widget_focus_mode_afc);
        } else {
            autoFocusText = getResources().getString(R.string.uxsdk_widget_focus_mode_auto);

        }

        makeSpannableString(autoFocusText, autoFocusTextColor, manualFocusTextColor);
    }

    private void makeSpannableString(String autoFocusText, int autoFocusColor, int manualFocusColor) {
        SpannableStringBuilder builder = new SpannableStringBuilder();

        SpannableString str1 = new SpannableString(autoFocusText);
        str1.setSpan(new ForegroundColorSpan(autoFocusColor), 0, str1.length(), 0);
        builder.append(str1);

        SpannableString str2 = new SpannableString(getResources().getString(R.string.uxsdk_widget_focus_mode_separator));
        str2.setSpan(new ForegroundColorSpan(Color.WHITE), 0, str2.length(), 0);
        builder.append(str2);

        SpannableString str3 = new SpannableString(getResources().getString(R.string.uxsdk_widget_focus_mode_manual));
        str3.setSpan(new ForegroundColorSpan(manualFocusColor), 0, str3.length(), 0);
        builder.append(str3);
        titleTextView.setText(builder, TextView.BufferType.SPANNABLE);
    }

    private void initAttributes(Context context, AttributeSet attrs) {
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.FocusModeWidget);
        if (!isInEditMode()) {
            setCameraIndex(SettingDefinitions.CameraIndex.find(typedArray.getInt(R.styleable.FocusModeWidget_uxsdk_cameraIndex, 0)));
        }
        activeColor = typedArray.getColor(R.styleable.FocusModeWidget_uxsdk_activeModeTextColor, getResources().getColor(R.color.uxsdk_green));
        inactiveColor = typedArray.getColor(R.styleable.FocusModeWidget_uxsdk_inactiveModeTextColor, getResources().getColor(R.color.uxsdk_white));
        Drawable background = typedArray.getDrawable(R.styleable.AutoExposureLockWidget_uxsdk_autoExposureUnlockDrawable);
        setTitleBackground(background);
        typedArray.recycle();
    }
    //endregion

    //region customizations

    /**
     * Set the index of camera to which the widget should react
     *
     * @param cameraIndex index of the camera.
     */
    public void setCameraIndex(@NonNull SettingDefinitions.CameraIndex cameraIndex) {
        if (!isInEditMode()) {
            widgetModel.setCameraIndex(cameraIndex);
        }
    }

    /**
     * Get the index of the camera to which the widget is reacting
     *
     * @return {@link SettingDefinitions.CameraIndex}
     */
    @NonNull
    public SettingDefinitions.CameraIndex getCameraIndex() {
        return widgetModel.getCameraIndex();
    }

    /**
     * Set active mode text color
     *
     * @param color color integer
     */
    public void setActiveModeTextColor(@ColorInt int color) {
        activeColor = color;
        checkAndUpdateUI();
    }

    /**
     * Get active mode text color
     *
     * @return color integer
     */
    @ColorInt
    public int getActiveModeTextColor() {
        return activeColor;
    }

    /**
     * Set in-active mode text color
     *
     * @param color color integer
     */
    public void setInactiveModeTextColor(@ColorInt int color) {
        inactiveColor = color;
        checkAndUpdateUI();
    }

    /**
     * Get in-active mode text color
     *
     * @return color integer
     */
    @ColorInt
    public int getInactiveModeTextColor() {
        return inactiveColor;
    }

    /**
     * Set background to title text
     *
     * @param resourceId resource id of background
     */
    public void setTitleBackground(@DrawableRes int resourceId) {
        setTitleBackground(getResources().getDrawable(resourceId));
    }

    /**
     * Set background to title text
     *
     * @param drawable Drawable to be used as background
     */
    public void setTitleBackground(@Nullable Drawable drawable) {
        titleTextView.setBackground(drawable);
    }

    /**
     * Get current background of title text
     *
     * @return Drawable resource of the background
     */
    @Nullable
    public Drawable getTitleBackground() {
        return titleTextView.getBackground();
    }

    /**
     * Sets the text size of the widget text
     *
     * @param textSize text size float value
     */
    public void setTitleTextSize(@Dimension float textSize) {
        titleTextView.setTextSize(textSize);
    }

    /**
     * Get current text size
     *
     * @return text size of the title
     */
    @Dimension
    public float getTitleTextSize() {
        return titleTextView.getTextSize();
    }

    //endregion
}

File: android-uxsdk-beta-cameracore/src/main/java/dji/ux/beta/cameracore/widget/focusmode/FocusModeWidget.kt
/*
 * Copyright (c) 2018-2021 DJI
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package dji.ux.beta.cameracore.widget.focusmode

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.drawable.Drawable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.style.ForegroundColorSpan
import android.util.AttributeSet
import android.view.View
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.Dimension
import androidx.annotation.DrawableRes
import androidx.core.content.res.use
import dji.common.camera.SettingsDefinitions.LensType
import dji.thirdparty.io.reactivex.Flowable
import dji.thirdparty.io.reactivex.functions.Action
import dji.thirdparty.io.reactivex.processors.PublishProcessor
import dji.ux.beta.cameracore.R
import dji.ux.beta.cameracore.widget.focusmode.FocusModeWidget.ModelState
import dji.ux.beta.cameracore.widget.focusmode.FocusModeWidgetModel.FocusModeState
import dji.ux.beta.core.base.DJISDKModel
import dji.ux.beta.core.base.SchedulerProvider
import dji.ux.beta.core.base.widget.FrameLayoutWidget
import dji.ux.beta.core.communication.GlobalPreferencesManager
import dji.ux.beta.core.communication.ObservableInMemoryKeyedStore
import dji.ux.beta.core.extension.*
import dji.ux.beta.core.util.SettingDefinitions.CameraIndex

private const val TAG = "FocusModeWidget"

/**
 * Widget will display the current focus mode of aircraft camera.
 * - MF text highlighted (in green) indicates focus mode is Manual Focus.
 * - AF text highlighted (in green) indicates focus mode is Auto Focus.
 * - AFC text highlighted (in green) indicates focus mode is Auto Focus Continuous.
 *
 *
 * Interaction:
 * Tapping will toggle between AF and MF mode.
 */
open class FocusModeWidget @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : FrameLayoutWidget<ModelState>(context, attrs, defStyleAttr), View.OnClickListener {

    //region Fields
    private val widgetModel: FocusModeWidgetModel by lazy {
        FocusModeWidgetModel(
                DJISDKModel.getInstance(),
                ObservableInMemoryKeyedStore.getInstance(),
                GlobalPreferencesManager.getInstance())
    }

    private val titleTextView: TextView = findViewById(R.id.text_view_camera_control_af)
    private val uiUpdateStateProcessor: PublishProcessor<UIState> = PublishProcessor.create()

    /**
     * Lens type for the widget
     */
    var lensType: LensType
        get() = widgetModel.lensType
        set(lensType) {
            if (!isInEditMode) {
                widgetModel.lensType = lensType
            }
        }

    /**
     * Camera key index for which this model should subscribe to.
     */
    var cameraIndex: CameraIndex
        get() = widgetModel.cameraIndex
        set(cameraIndex) {
            if (!isInEditMode) {
                widgetModel.cameraIndex = cameraIndex
            }
        }

    /**
     * Color for the text of active mode
     */
    @ColorInt
    var activeModeTextColor = getColor(R.color.uxsdk_green)
        set(value) {
            field = value
            checkAndUpdateUI()
        }

    /**
     * Color for the text of inactive mode
     */
    @ColorInt
    var inactiveModeTextColor = getColor(R.color.uxsdk_white)
        set(value) {
            field = value
            checkAndUpdateUI()
        }

    /**
     * The color of the text when the product is disconnected
     */
    @get:ColorInt
    var disconnectedStateTextColor: Int = getColor(R.color.uxsdk_gray_58)
        set(@ColorInt value) {
            field = value
            checkAndUpdateUI()
        }


    /**
     * Text size of the widget text
     *
     */
    @get:Dimension
    var titleTextSize: Float
        get() = titleTextView.textSize
        set(textSize) {
            titleTextView.textSize = textSize
        }

    /**
     * background to title text
     */
    var titleBackground: Drawable?
        get() = titleTextView.background
        set(value) {
            titleTextView.background = value
        }
    //endregion

    //region Lifecycle

    init {
        background = background ?: getDrawable(R.drawable.uxsdk_background_black_rectangle)
        attrs?.let { initAttributes(context, it) }
    }

    override fun initView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
        inflate(context, R.layout.uxsdk_widget_focus_mode_switch, this)
        setOnClickListener(this)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (!isInEditMode) {
            widgetModel.setup()
        }
    }

    override fun onDetachedFromWindow() {
        if (!isInEditMode) {
            widgetModel.cleanup()
        }
        super.onDetachedFromWindow()
    }

    override fun reactToModelChanges() {
        addReaction(widgetModel.productConnection
                .observeOn(SchedulerProvider.ui())
                .subscribe { widgetStateDataProcessor.onNext(ModelState.ProductConnected(it)) })
        addReaction(widgetModel.focusModeState
                .observeOn(SchedulerProvider.ui())
                .subscribe { updateUI(it) })
    }

    override fun getIdealDimensionRatioString(): String {
        return resources.getString(R.string.uxsdk_widget_default_ratio)
    }

    override fun onClick(v: View) {
        if (isEnabled) {
            uiUpdateStateProcessor.onNext(UIState.WidgetClicked)
            addDisposable(widgetModel.toggleFocusMode()
                    .observeOn(SchedulerProvider.ui())
                    .subscribe(Action {}, logErrorConsumer(TAG, "switch focus mode: ")))
        }
    }

    //endregion

    //region private helpers
    private fun checkAndUpdateUI() {
        if (!isInEditMode) {
            addDisposable(widgetModel.focusModeState.firstOrError()
                    .observeOn(SchedulerProvider.ui())
                    .subscribe({ updateUI(it) },
                            { logErrorConsumer(TAG, "Update UI ") }))
        }
    }

    private fun updateUI(focusModeState: FocusModeState) {
        widgetStateDataProcessor.onNext(ModelState.FocusModeStateUpdated(focusModeState))
        when (focusModeState) {
            FocusModeState.ProductDisconnected,
            FocusModeState.CameraDisconnected,
            FocusModeState.NotSupported -> updateDisconnectedState()
            is FocusModeState.ManualFocus -> updateFocusModeString(getAutoFocusString(focusModeState.isAFCEnabled),
                    inactiveModeTextColor,
                    activeModeTextColor)
            is FocusModeState.AutoFocus -> updateFocusModeString(getAutoFocusString(focusModeState.isAFCEnabled),
                    activeModeTextColor,
                    inactiveModeTextColor)
        }
    }

    private fun updateDisconnectedState() {
        titleTextView.setTextColor(disconnectedStateTextColor)
        isEnabled = false
        hide()
    }

    private fun getAutoFocusString(isAFCEnabled: Boolean): String {
        return if (isAFCEnabled) {
            getString(R.string.uxsdk_widget_focus_mode_afc)
        } else {
            getString(R.string.uxsdk_widget_focus_mode_auto)
        }
    }

    private fun updateFocusModeString(autoFocusText: String, autoFocusColor: Int, manualFocusColor: Int) {
        val builder = SpannableStringBuilder()
        val str1 = SpannableString(autoFocusText)
        str1.setSpan(ForegroundColorSpan(autoFocusColor), 0, str1.length, 0)
        builder.append(str1)
        val str2 = SpannableString(resources.getString(R.string.uxsdk_widget_focus_mode_separator))
        str2.setSpan(ForegroundColorSpan(inactiveModeTextColor), 0, str2.length, 0)
        builder.append(str2)
        val str3 = SpannableString(resources.getString(R.string.uxsdk_widget_focus_mode_manual))
        str3.setSpan(ForegroundColorSpan(manualFocusColor), 0, str3.length, 0)
        builder.append(str3)
        titleTextView.setText(builder, TextView.BufferType.SPANNABLE)
        isEnabled = true
        show()
    }

    @SuppressLint("Recycle")
    private fun initAttributes(context: Context, attrs: AttributeSet) {
        context.obtainStyledAttributes(attrs, R.styleable.FocusModeWidget).use { typedArray ->
            typedArray.getIntegerAndUse(R.styleable.FocusModeWidget_uxsdk_cameraIndex) {
                cameraIndex = CameraIndex.find(it)
            }
            typedArray.getIntegerAndUse(R.styleable.FocusModeWidget_uxsdk_lensType) {
                lensType = LensType.find(it)
            }
            typedArray.getColorAndUse(R.styleable.FocusModeWidget_uxsdk_activeModeTextColor) {
                activeModeTextColor = it
            }
            typedArray.getColorAndUse(R.styleable.FocusModeWidget_uxsdk_inactiveModeTextColor) {
                inactiveModeTextColor = it
            }
            typedArray.getColorAndUse(R.styleable.FocusModeWidget_uxsdk_disconnectedStateTextColor) {
                disconnectedStateTextColor = it
            }
            typedArray.getDimensionAndUse(R.styleable.FocusModeWidget_uxsdk_widgetTitleTextSize) {
                titleTextSize = it
            }
            typedArray.getDrawableAndUse(R.styleable.FocusModeWidget_uxsdk_widgetTitleBackground) {
                titleBackground = it
            }
        }
    }
    //endregion

    //region customizations

    /**
     * Set background to title text
     *
     * @param resourceId resource id of background
     */
    fun setTitleBackground(@DrawableRes resourceId: Int) {
        titleBackground = getDrawable(resourceId)
    }
    //endregion

    //region hooks
    /**
     * Get the [UIState] updates
     */
    fun getUIStateUpdates(): Flowable<UIState> {
        return uiUpdateStateProcessor.onBackpressureBuffer()
    }

    /**
     * Class defines the widget state updates
     */
    sealed class ModelState {
        /**
         * Product connection update
         */
        data class ProductConnected(val isConnected: Boolean) : ModelState()

        /**
         * Focus mode state update
         */
        data class FocusModeStateUpdated(val focusModeState: FocusModeState) : ModelState()
    }

    /**
     * Class defines the UI state updates
     */
    sealed class UIState {
        /**
         * Widget click update
         */
        object WidgetClicked : UIState()
    }
    //endregion
}